File: cdbfapi-ios/cdbfapi-ios-example/ViewController.m
//
//  ViewController.m
//  cdbfapi-ios-example
//
//  Created by Sergey Chehuta on 17/09/2017.
//  Copyright © 2017 WhiteTown. All rights reserved.
//

#import "ViewController.h"
#import <cdbfapi/cdbfapi.h>

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *tableViewS;
@property (weak, nonatomic) IBOutlet UITableView *tableViewD;

@property (strong, nonatomic) cdbfapi *dbf;

@end

@implementation ViewController

#define cellID @"Cell"

- (void)viewDidLoad
{
    [super viewDidLoad];

    [self.tableViewD registerClass:[UITableViewCell class] forCellReuseIdentifier:cellID];
    [self.tableViewS registerClass:[UITableViewCell class] forCellReuseIdentifier:cellID];
}

- (IBAction)openDBF:(id)sender
{
    NSString *filename = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:@"example.dbf"];

    self.dbf = [cdbfapi new];
    [self.dbf openDBFfile:filename];

    [self.dbf setDateFormat:@"dmy"];

    [self.tableViewS reloadData];
    [self.tableViewD reloadData];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    if (!self.dbf) return 0;

    if ([tableView isEqual:self.tableViewS])
        return [self.dbf fieldCount];

    if ([tableView isEqual:self.tableViewD])
        return [self.dbf recCount];

    return 0;
}

- (UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellID];

    if ([tableView isEqual:self.tableViewS])
    {
        cell.textLabel.text = [NSString stringWithFormat:@"%@ %c(%ld.%ld)",
                              [self.dbf fieldName:indexPath.row],
                              [self.dbf fieldType:indexPath.row],
                              (long)[self.dbf fieldLength:indexPath.row],
                              (long)[self.dbf fieldDecimal:indexPath.row]
                              ];
    }

    if ([tableView isEqual:self.tableViewD])
    {
        NSMutableArray *data = [NSMutableArray new];

        [self.dbf getRecord:indexPath.row];

        for(NSInteger field = 0; field < [self.dbf fieldCount]; field++)
        {
            [data addObject: [[self.dbf getString:field] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] ? : @""];
            /*
            if ([self.dbf fieldType:field] == 'D')
            {
                NSDate *date = [self.dbf getDateTime:field];
                NSString *dateString = [NSDateFormatter localizedStringFromDate:date
                                                                      dateStyle:NSDateFormatterShortStyle
                                                                      timeStyle:NSDateFormatterFullStyle];
                NSLog(@"%@",dateString);
            }
             */
        }
        cell.textLabel.text = [data componentsJoinedByString:@", "];
    }

    return cell;
}

@end

File: cdbfapi-ios/cdbfapi-ios-swift/ViewController.swift
//
//  ViewController.swift
//  cdbfapi-ios-swift
//
//  Created by Sergey Chehuta on 26/09/2017.
//  Copyright © 2017 WhiteTown. All rights reserved.
//

import UIKit
import cdbfapi

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    @IBOutlet weak var tableViewS: UITableView!
    @IBOutlet weak var tableViewD: UITableView!

    var dbf = cdbfapi()

    let cellID = "Cell"

    override func viewDidLoad() {
        super.viewDidLoad()

        self.tableViewS.register(UITableViewCell.self, forCellReuseIdentifier: cellID)
        self.tableViewD.register(UITableViewCell.self, forCellReuseIdentifier: cellID)
    }

    @IBAction func btnTap(_ sender: Any)
    {
        let filename = Bundle.main.resourcePath?.appending("/example.dbf")

        self.dbf = cdbfapi()
        self.dbf.openDBFfile(filename!)

        self.dbf.setDateFormat("dmy")

        self.tableViewS.reloadData()
        self.tableViewD.reloadData()
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if tableView == tableViewS {
            return self.dbf.fieldCount()
        } else {
            return self.dbf.recCount()
        }
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        let cell = tableView.dequeueReusableCell(withIdentifier: cellID)

        if tableView == tableViewS {
            cell?.textLabel?.text = String.init(format: "%@ %c(%ld.%ld)",
                                    self.dbf.fieldName(indexPath.row) ?? "-",
                                    self.dbf.fieldType(indexPath.row),
                                    self.dbf.fieldLength(indexPath.row),
                                    self.dbf.fieldDecimal(indexPath.row));

        } else {
            self.dbf.getRecord(indexPath.row)
            var data = [String]()
            for field in 0...self.dbf.fieldCount()-1
            {
                data.append((self.dbf.getString(field) ?? "").trimmingCharacters(in: .whitespacesAndNewlines) )
            }

            cell?.textLabel?.text = data.joined(separator: ", ")
        }

        return cell!
    }


}

